lib/spack/spack/hooks/check_dynamic_linking_elf.py

```diff
@@ -0,0 +1,172 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+import struct
+
+import llnl.util.tty as tty
+from llnl.util.lang import elide_list
+
+import spack.error
+import spack.object_file.elf as elf
+
+skip_list = set([
+    # kernel
+    'linux-vdso',
+    # glibc
+    'ld-linux-x86-64',
+    'libc',
+    'libdl',
+    'libgcc_s',
+    'libm',
+    'libmemusage',
+    'libmvec',
+    'libnsl',
+    'libnss_compat',
+    'libnss_db',
+    'libnss_dns',
+    'libnss_files',
+    'libnss_hesiod',
+    'libpcprofile',
+    'libpthread',
+    'libresolv',
+    'librt',
+    'libSegFault',
+    'libthread_db',
+    'libutil',
+    # gcc
+    'libasan',
+    'libatomic',
+    'libcc1',
+    'libgcc_s',
+    'libgfortran',
+    'libgomp',
+    'libitm',
+    'liblsan',
+    'libquadmath',
+    'libssp',
+    'libstdc++',
+    'libtsan',
+    'libubsan',
+])
```

Comment:
Curating this list will likely be an ongoing (and initially-frequent) process. I'm wondering if we can circumvent this curation by searching for libraries that meet the following criteria:

* The library is present in a dependency
* We are getting it from somewhere else

This would involve perhaps creating a DB of all library files in all dependencies (which would be useful for other purposes e.g. better error messages when an installation fails because of missing symbols).

Reply:
Makes sense, but it does not help with errors like #28114: if multiple libraries are supported and the build links to the first it finds on the system (e.g. berkely-db/intel-mkl on the system, whereas gdbm/openblas was the spack dependency).

Reply:
I'm also thinking that something stateless is better than yet another cache

Reply:
Also we have to deal with stuff like MKL which:

- installs libs with the same filename, soname, arch in different directories depending on what interface the user needs (32 vs 64 bit integers on a 64 bit arch)
- installs a copy of Qt for their updated/installer

So "the library is present in a dependency" is not a guarantee that it's meant to be exposed to the user, which is the reason I went for `spec[x].libs.directories` as candidate search paths.

Reply:
Notice that the MKL issue also means we can't just randomly inject rpaths to MKL in the first place.

Reply:
And... turns out `.libs` can't be trusted either. For mkl it returns glibc libs :weary:. I don't see how to make any progress here.

Reply:
> Makes sense, but it does not help with errors like #28114: if multiple libraries are supported and the build links to the first it finds on the system (e.g. berkely-db/intel-mkl on the system, whereas gdbm/openblas was the spack dependency).

Sorry I don't think I understand: wouldn't running `ldd` for example give the locations of all found dependency libraries? In this case for example I think the error we would detect is:

* `berkeley-db` installed by Spack has a library `X`
* `ldd` reports finding `X`, but in a system path
* Now we know there's an issue, because we should be getting `X` from the Spack package

One place where my suggestion wouldn't work is if Spack's Python package did not mention `berkeley-db` as a transitive dependency (in that case Spack may not install it, and hence may not know that `X` is provided). It looks like that might happen for example if one uses an external `openssl`.

> So "the library is present in a dependency" is not a guarantee that it's meant to be exposed to the user, 

`.libs` should not return such libraries, so we can also harvest the names from that (`.libs` is a `LibraryList` object which can give you `.directories` or `.filenames` - in the latter case the individual library files).

> I'm also thinking that something stateless is better than yet another cache

I agree that caches have plenty of issues, although IMO in this case the benefits are worthwhile. If we can find a way to limit the complaints that end users get then this solution could be the right way to go. I think curating `skip_list` will take significant work:

* I assume each different compiler will have libraries (right now you have ones associated with GCC, but I assume there will need to be at least one entry for Intel, `nvhpc`, `go`, etc.
  * Also, whether or not it's "OK" to find some of these on the system depends on whether we've installed a compiler with Spack
* I assume there may be minor differences for different operating systems (MacOS, and perhaps other Linux distributions)

That doesn't actually seem so bad to manage, now that I list it out. I wonder though if I'm missing something.

Perhaps this can be mitigated by only optionally running this post-install hook. I think that for example it would always make sense to run this for all CI jobs on all systems.